# Battle of the Neighbourhoods - Toronto

Author:Rajeswari

In [4]:
import pandas as pd

In [5]:

import requests

In [7]:

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [8]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [9]:
len(wiki_data), type(wiki_data)

(3, list)

In [10]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [11]:

df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [14]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [15]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [16]:
df.shape

(103, 3)

##Answer to Question 1: We have 103 rows and 3 columns

### Question2

In [73]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [18]:
import geocoder # import geocoder

In [19]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [20]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


#datatype

In [22]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [23]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [24]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [25]:
combined_data.shape

(103, 5)

### Question 3

In [26]:
from geopy.geocoders import Nominatim

In [27]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


#### Initializing Foursquare API credentials

In [30]:
CLIENT_ID = 'MFOBXEURQL3A2FKHFLC0B5MQARXCVRLZMUSGRYFUQ2L2ELVF' 
CLIENT_SECRET = 'MDO1KEOXUYUQQ0WDTOBG3B3XD3USUSFQ3Z3Y3MJHUWZWW2TH'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MFOBXEURQL3A2FKHFLC0B5MQARXCVRLZMUSGRYFUQ2L2ELVF
CLIENT_SECRET:MDO1KEOXUYUQQ0WDTOBG3B3XD3USUSFQ3Z3Y3MJHUWZWW2TH


In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        #create venues in toronto    
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

#Collecting the venues in Toronto for each Neighbourhood

In [32]:

venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [33]:
venues_in_toronto.shape

(1338, 5)

In [34]:
#we have 1338 records and 5 columns

In [35]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                               Venue              Venue Category  
0                    Brookbanks Park                        Park  
1                Towns On The Ravine                       Hotel  
2                      Variety Store           Food & Drink Shop  
3  Corrosion Service Company Limited  Construction & Landscaping  
4             Victoria Village Arena                Hockey Arena

In [36]:
#based on neighbourhood

In [37]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                             Parkwoods   
4                                      Victoria Village   
...                                                 ...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   
1325  Mimico NW, The Queensway West, South of Bloor,...   
1326  Mimico NW, The Queensway West, South of Bloor,...   
1327  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.753259               -79.329656   
3                  43.753259               -79.329656   
4                  43.725882               -79.315572   
...                      ...                      ...   
1323               43.628841               -79.520999   
1324               43.628841               -79.520999   
1325               43.628841               -79.520999   
1326               43.628841               -79.520999   
1327               43.628841               -79.520999   

                                  Venue              Venue Category  
0                       Brookbanks Park                        Park  
1                   Towns On The Ravine                       Hotel  
2                         Variety Store           Food & Drink Shop  
3     Corrosion Service Company Limited  Construction & Landscaping  
4                Victoria Village Arena                Hockey Arena  
...                                 ...                         ...  
1323                         Wingporium                 Wings Joint  
1324                   South St. Burger                Burger Joint  
1325                          Dollarama              Discount Store  
1326                     Healthy Planet             Supplement Shop  
1327                           7-Eleven           Convenience Store  

[416 rows x 5 columns]

In [38]:
#416 record#

In [39]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.733283               -79.419750   
Yoga Studio                      43.715383               -79.321558   

                                               Venue  
Venue Category                                        
Accessories Store                Ardene Shoes Outlet  
Airport              Toronto Downsview Airport (YZD)  
Airport Food Court                 Billy Bishop Café  
Airport Lounge                         Porter Lounge  
Airport Service     Porter Airlines Check-In Counter  
...                                              ...  
Warehouse Store                               Costco  
Wine Bar                             Paris Paris Bar  
Wings Joint                               Wingporium  
Women's Store                          Want Boutique  
Yoga Studio                                YogaSpace  

[235 rows x 4 columns]

In [40]:
#235 venue categories

In [41]:
#hot encodings
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1333                  0        0                   0               0   
1334                  0        0                   0               0   
1335                  0        0                   0               0   
1336                  0        0                   0               0   
1337                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1333                0                 0                    0             0   
1334                0                 0                    0             0   
1335                0                 0                    0             0   
1336                0                 0                    0             0   
1337                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Turkish Restaurant  \
0            0            0  ...                   0   
1            0            0  ...                   0   
2            0            0  ...                   0   
3            0            0  ...                   0   
4            0            0  ...                   0   
...        ...          ...  ...                 ...   
1333         0            0  ...                   0   
1334         0            0  ...                   0   
1335         0            0  ...                   0   
1336         0            0  ...                   0   
1337         0            0  ...                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1333                              0                 0            0   
1334                              0                 0            0   
1335                              0                 0            0   
1336                              0                 0            0   
1337                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1333                      0                0         0            0   
1334                      0                0         0            0   
1335                      0                0       

In [42]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3         Parkwoods                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Turkish Restaurant  \
0             0         0  ...                   0   
1             0         0  ...                   0   
2             0         0  ...                   0   
3             0         0  ...                   0   
4             0         0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 236 columns]

In [43]:
#calculate the mean value for each venue categories
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0               0.0             0.000000           0.0       0.0  ...   
1               0.0             0.000000           0.0       0.0  ...   
2               0.0             0.000000           0.0       0.0  ...   
3               0.0             0.000000           0.0       0.0  ...   
4               0.0             0.038462           0.0       0.0  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0       0.000000          0.0  
1       0.000000          0.0  
2       0.000000          0.0  
3       0.000000          0.0  
4       0.038462          0.0  

[5 rows x 236 columns]

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
import numpy as np

In [46]:
#top 10 cluster
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot        Clothing Store          Skating Rink   
1              Pharmacy        Sandwich Place                  Pool   
2                  Bank                 Diner   Fried Chicken Joint   
3    Chinese Restaurant                  Bank   Japanese Restaurant   
4    Italian Restaurant           Coffee Shop     Indian Restaurant   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                Lounge                Yoga Studio          Dessert Shop   
1                   Pub                Coffee Shop                   Gym   
2      Sushi Restaurant  Middle Eastern Restaurant        Ice Cream Shop   
3      Department Store                Escape Room     Electronics Store   
4                   Pub                       Café               Butcher   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1                      Gay Bar              Coworking Space   
2                   Restaurant            Mobile Phone Shop   
3  Eastern European Restaurant                    Drugstore   
4             Sushi Restaurant                 Liquor Store   

  10th Most Common Venue  
0              Drugstore  
1         Discount Store  
2          Deli / Bodega  
3             Donut Shop  
4             Hobby Shop

In [47]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [48]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [49]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       4, 1, 4, 1, 1, 2, 4, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4,
       1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       4, 1, 1, 1, 1, 1, 4], dtype=int32)

In [50]:
#lets add clustering label column
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [51]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels   1st Most Common Venue  \
0  43.753259 -79.329656             1.0                    Park   
1  43.725882 -79.315572             1.0             Pizza Place   
2  43.654260 -79.360636             1.0             Coffee Shop   
3  43.718518 -79.464763             1.0  Furniture / Home Store   
4  43.662301 -79.389494             1.0             Coffee Shop   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0  Construction & Landscaping      Food & Drink Shop                 Hotel   
1                 Coffee Shop  Portuguese Restaurant     French Restaurant   
2                        Park                 Bakery        Breakfast Spot   
3              Clothing Store      Accessories Store           Coffee Shop   
4           College Cafeteria                  Diner      Sushi Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Yoga Studio         Deli / Bodega  Eastern European Restaurant   
1          Hockey Arena                 Diner                Deli / Bodega   
2                  Café               Theater                  Yoga Studio   
3    Miscellaneous Shop              Boutique                  Event Space   
4           Yoga Studio                  Café               Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Drugstore            Donut Shop                Dog Run  
1       Department Store          Dessert Shop     Dim Sum Restaurant  
2       Greek Restaurant                   Pub  Performing Arts Venue  
3  Vietnamese Restaurant         Women's Store           Dance Studio  
4          Smoothie Shop   Distribution Center          Burrito Place

In [52]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [53]:
#Plotting the clusters

In [61]:
import matplotlib.cm as cm
import matplotlib.colors as colors
!conda install -c conda-forge folium=0.5.0 

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.7-main

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |           1_llvm           5 KB  conda-forge
    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge
    _pytorch_select-0.2        |            gpu_0           2 KB
    absl-py-0.11.0             |   py37h89c1867_0         168 KB  conda-forge
    aiohttp-3.7.4              |   py37h5e8e339_0         632 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
   

python-dateutil-2.8. | 220 KB    | ##################################### | 100% 
keras-applications-1 | 30 KB     | ##################################### | 100% 
decorator-4.4.2      | 11 KB     | ##################################### | 100% 
pyasn1-modules-0.2.7 | 60 KB     | ##################################### | 100% 
scipy-1.5.3          | 18.5 MB   | ##################################### | 100% 
toolz-0.11.1         | 46 KB     | ##################################### | 100% 
tqdm-4.58.0          | 76 KB     | ##################################### | 100% 
beautifulsoup4-4.9.3 | 86 KB     | ##################################### | 100% 
grpcio-1.36.1        | 2.1 MB    | ##################################### | 100% 
libgfortran4-7.5.0   | 1.3 MB    | ##################################### | 100% 
unixodbc-2.3.9       | 293 KB    | ##################################### | 100% 
scikit-learn-0.24.1  | 7.5 MB    | ##################################### | 100% 
mock-4.0.3           | 51 KB

fontconfig-2.13.1    | 344 KB    | ##################################### | 100% 
jedi-0.18.0          | 923 KB    | ##################################### | 100% 
libevent-2.1.10      | 1.1 MB    | ##################################### | 100% 
kiwisolver-1.3.1     | 78 KB     | ##################################### | 100% 
s3transfer-0.3.4     | 51 KB     | ##################################### | 100% 
jinja2-2.11.3        | 93 KB     | ##################################### | 100% 
keyring-18.0.0       | 50 KB     | ##################################### | 100% 
libblas-3.9.0        | 12 KB     | ##################################### | 100% 
seaborn-0.11.1       | 4 KB      | ##################################### | 100% 
click-7.1.2          | 64 KB     | ##################################### | 100% 
docutils-0.16        | 739 KB    | ##################################### | 100% 
keras-preprocessing- | 34 KB     | ##################################### | 100% 
jupyterlab_pygments- | 8 KB 

mkl_random-1.2.0     | 313 KB    | ##################################### | 100% 
pyqt-impl-5.12.3     | 5.9 MB    | ##################################### | 100% 
vincent-0.4.4        | 28 KB     | ##################################### | 100% 
werkzeug-1.0.1       | 239 KB    | ##################################### | 100% 
pyjwt-2.0.1          | 17 KB     | ##################################### | 100% 
markdown-3.3.4       | 67 KB     | ##################################### | 100% 
async-timeout-3.0.1  | 11 KB     | ##################################### | 100% 
appdirs-1.4.4        | 13 KB     | ##################################### | 100% 
ca-certificates-2020 | 137 KB    | ##################################### | 100% 
libgcc-ng-9.3.0      | 7.8 MB    | ##################################### | 100% 
defusedxml-0.6.0     | 22 KB     | ##################################### | 100% 
libxcb-1.13          | 395 KB    | ##################################### | 100% 
icu-68.1             | 13.0 

done


In [75]:
import pandas as pd


map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

In [64]:
#cluster1 verification
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             0.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Event Space           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

In [65]:
#cluster2 verification
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
0          North York             1.0                    Park   
1          North York             1.0             Pizza Place   
2    Downtown Toronto             1.0             Coffee Shop   
3          North York             1.0  Furniture / Home Store   
4    Downtown Toronto             1.0             Coffee Shop   
..                ...             ...                     ...   
98          Etobicoke             1.0                   River   
99   Downtown Toronto             1.0        Sushi Restaurant   
100      East Toronto             1.0             Yoga Studio   
101         Etobicoke             1.0          Baseball Field   
102         Etobicoke             1.0             Social Club   

          2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0    Construction & Landscaping      Food & Drink Shop                 Hotel   
1                   Coffee Shop  Portuguese Restaurant     French Restaurant   
2                          Park                 Bakery        Breakfast Spot   
3                Clothing Store      Accessories Store           Coffee Shop   
4             College Cafeteria                  Diner      Sushi Restaurant   
..                          ...                    ...                   ...   
98                   Smoke Shop                   Pool           Curling Ice   
99                  Coffee Shop                   Café              Beer Bar   
100                  Smoke Shop          Auto Workshop               Brewery   
101                Home Service            Yoga Studio      Department Store   
102           Convenience Store         Sandwich Place            Kids Store   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0             Yoga Studio         Deli / Bodega  Eastern European Restaurant   
1            Hockey Arena                 Diner                Deli / Bodega   
2                    Café               Theater                  Yoga Studio   
3      Miscellaneous Shop              Boutique                  Event Space   
4             Yoga Studio                  Café               Sandwich Place   
..                    ...                   ...                          ...   
98              Drugstore            Donut Shop                      Dog Run   
99       Ramen Restaurant   Martial Arts School                   Steakhouse   
100         Burrito Place               Butcher                   Comic Shop   
101           Escape Room     Electronics Store  Eastern European Restaurant   
102  Fast Food Restaurant          Burger Joint                Burrito Place   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Drugstore            Donut Shop                Dog Run  
1         Department Store          Dessert Shop     Dim Sum Restaurant  
2         Greek Restaurant                   Pub  Performing Arts Venue  
3    Vietnamese Restaurant         Women's Store           Dance Studio  
4            Smoothie Shop   Distribution Center          Burrito Place  
..                     ...                   ...                    ...  
98     Distribution Center        Discount Store                  Diner  
99       Indian Restaurant  Ethiopian Restaurant            Escape Room  
100         Farmers Market  Fast Food Restaurant                 Garden  
101              Drugstore            Donut Shop                Dog Run  
102         Discount Store       Supplement Shop                 Bakery  

[87 rows x 12 columns]

In [66]:
#Cluster 3 verification
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

In [67]:
#cluster 4 verification

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             3.0            Playground          Dance Studio   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
32     Electronics Store  Eastern European Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
32        Discount Store                  Diner

In [ ]:
#cluster5 verification
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]